# Portfolio assignment 20
**Train a perceptron to predict the number of the MNIST dataset.:**

- Fit a Perceptron model (keep hidden_layer_sizes=1) using the images in de fetch openml dataset.
- Change the learning_rate and max_iter to find the 'right fit'.
- Use your perceptron to make predictions for both the train and test set.

In [8]:
from sklearn.datasets import fetch_openml
from sklearn.linear_model import Perceptron
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from sklearn.exceptions import ConvergenceWarning

# Ignore the ConvergenceWarning
warnings.filterwarnings("ignore", category=ConvergenceWarning)


# Load the MNIST dataset
mnist = fetch_openml('mnist_784', as_frame=False)
X = mnist.data
y = mnist.target.astype(int)
print(X.shape)  # (70000, 784)
print(y.shape)  # (70000,)

(70000, 784)
(70000,)


- Fine-tune the learning rate and epochs (max iterations).
- Calculate the accuracy for both the train set predictions and test set predictions, what happens per iteration?
- Is the accurracy different? Did you expect this difference?

---

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.14, random_state=42)

In [6]:
# Initieel model
perceptron = Perceptron(max_iter=50, eta0=1.0, random_state=42)
perceptron.fit(X_train, y_train)

# Voorspellingen
y_pred_train = perceptron.predict(X_train)
y_pred_test = perceptron.predict(X_test)

# Accuratesse berekenen
acc_train = accuracy_score(y_train, y_pred_train)
acc_test = accuracy_score(y_test, y_pred_test)

print(f"Train accuracy: {acc_train:.2f}")
print(f"Test accuracy: {acc_test:.2f}")

Train accuracy: 0.89
Test accuracy: 0.89


In [10]:
perceptron = Perceptron(max_iter=20, eta0=1.0, random_state=42)
perceptron.fit(X_train, y_train)

y_pred_train = perceptron.predict(X_train)
y_pred_test = perceptron.predict(X_test)

acc_train = accuracy_score(y_train, y_pred_train)
acc_test = accuracy_score(y_test, y_pred_test)

print(f"Train accuracy: {acc_train:.2f}")
print(f"Test accuracy: {acc_test:.2f}")

Train accuracy: 0.90
Test accuracy: 0.89


In [11]:
perceptron = Perceptron(max_iter=100, eta0=0.5, random_state=42)
perceptron.fit(X_train, y_train)

y_pred_train = perceptron.predict(X_train)
y_pred_test = perceptron.predict(X_test)

acc_train = accuracy_score(y_train, y_pred_train)
acc_test = accuracy_score(y_test, y_pred_test)

print(f"Train accuracy: {acc_train:.2f}")
print(f"Test accuracy: {acc_test:.2f}")

Train accuracy: 0.89
Test accuracy: 0.89


In [12]:
perceptron = Perceptron(max_iter=150, eta0=0.3, random_state=42)
perceptron.fit(X_train, y_train)

y_pred_train = perceptron.predict(X_train)
y_pred_test = perceptron.predict(X_test)

acc_train = accuracy_score(y_train, y_pred_train)
acc_test = accuracy_score(y_test, y_pred_test)

print(f"Train accuracy: {acc_train:.2f}")
print(f"Test accuracy: {acc_test:.2f}")

Train accuracy: 0.89
Test accuracy: 0.89


### Findings:
**Is de accuracy verschillend? Verwacht?**
- niet echt, train accuracy is heel soms iets hoger dan test accuracy, Het verschil in accuracy tussen train en test is klein, dus geen overfitting.
- Andere combinaties van max_iter en eta0 gaven dus vergelijkbare resultaten (rond 89%)
- Dit is normaal: het model leert specifieke patronen uit de trainingsset die niet altijd exact in de testset terugkomen.
- Te veel epochs kan overfitting veroorzaken (te hoge train accuracy(>90%), lagere test accuracy(<70%)).

**Begrijp je wat de Perceptron doet?**
- De Perceptron leert lineaire scheidingsvlakken in 784-dimensionale ruimte (1 per cijfer).
- Geen hidden layer betekent alleen lineaire beslissingen, geen complexere patronen (zoals in deep learning).
- MNIST is deels lineair scheidbaar, dus met genoeg iteraties en juiste learning rate lukt het goed.